### PR TITLE
Refactor ExpressionMap into ExpressionCollection

### PR DESCRIPTION
**Pull request by a046**
Next part of #171, supporting multiple expressions with the same name. Goal is to allow multiple KeySelectorAsc/KeySelectorDesc to be added which can then be used (in order) for a multi-key sort.

The approach I've taken with NamedLambdaExpression/NamedAggregateExpression I think isn't ideal for the CustomAggregator API (+ some internal bits) but I couldn't really think of a nicer way to do it.

### CODE REVIEW

File: src/NRules/NRules/Aggregators/NamedAggregateExpression.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+﻿using NRules.RuleModel;
+
+namespace NRules.Aggregators
+{
+    /// <summary>
+    /// Aggregate Expression with a name used in constructing/building an aggregator.
+    /// </summary>
+    public class NamedAggregateExpression : INamedExpression<IAggregateExpression>
```
Comment (snikolayev) [inline src/NRules/NRules/Aggregators/NamedAggregateExpression.cs]: I don't think this is needed. I would ditch it, and just add Name to IAggregateExpression

File: src/NRules/NRules.RuleModel/Builders/AggregateBuilder.cs [hunk 2]
```diff
@@ -11,7 +11,7 @@ public class AggregateBuilder : RuleElementBuilder, IBuilder<AggregateElement>
     {
         private string _name;
         private Type _resultType;
-        private readonly Dictionary<string, LambdaExpression> _expressions = new Dictionary<string, LambdaExpression>();
```
Comment (snikolayev) [inline src/NRules/NRules.RuleModel/Builders/AggregateBuilder.cs:14]: The idea was to use primitives in builders and static factories. So, I was thinking to use `KVP<string, LambdaExpression>` here and in the static factories in Element.cs.
What do you think?

File: src/NRules/NRules.RuleModel/Builders/Element.cs [hunk 3]
```diff
@@ -409,7 +409,7 @@ public static BindingElement Binding(LambdaExpression expression)
         /// <param name="expressions">Expressions used to construct aggregates from individual facts.</param>
         /// <param name="source">Pattern that matches facts for aggregation.</param>
         /// <returns>Created element.</returns>
-        public static AggregateElement Aggregate(Type resultType, string name, IDictionary<string, LambdaExpression> expressions, PatternElement source)
```
Comment (snikolayev) [inline src/NRules/NRules.RuleModel/Builders/Element.cs:412]: See my comment on AggregateBuilder, was thinking about using primitives here, like `KVP<string, LambdaExpression>`

File: src/NRules/NRules.RuleModel/NamedExpressionElement.cs [hunk 4]
```diff
@@ -7,10 +7,10 @@ namespace NRules.RuleModel
     /// </summary>
     public class NamedExpressionElement : ExpressionElement
     {
-        internal NamedExpressionElement(string name, LambdaExpression expression) 
-            : base(expression)
+        internal NamedExpressionElement(NamedLambdaExpression namedLambdaExpression)
```
Comment (snikolayev) [inline src/NRules/NRules.RuleModel/NamedExpressionElement.cs]: It's much better to pass name and expression separately here, to not couple to named expression representation. So, I would revert the change in this file.

File: src/NRules/NRules/Aggregators/NamedAggregateExpression.cs (hunk 1 above)
Comment (snikolayev) [reply src/NRules/NRules/Aggregators/NamedAggregateExpression.cs]: And `INamedExpression` also becomes unnecessary

File: src/NRules/NRules.RuleModel/NamedLambdaExpression.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+
+namespace NRules.RuleModel
+{
+    /// <summary>
+    /// Lambda Expression with a name used in constructing/building an aggregator.
+    /// </summary>
+    public class NamedLambdaExpression : INamedExpression<LambdaExpression>
```
Comment (snikolayev) [inline src/NRules/NRules.RuleModel/NamedLambdaExpression.cs]: Overall, I was thinking to use the following constructs:
* Primitives, like `KVP<string, LambdaExpression>` in builders and static factories.
* ExpressionCollection and NamedExpressionElement in the canonical model
* IAggregateExpression in the executable model (just add a Name to it, see my other comment)

Then, this NamedLambdaExpression is not needed, likewise INamedExpression and NamedAggregateExpression. Want to reduce coupling between layers as much as possible.

**Review by snikolayev — COMMENTED** [review]
This is a tough one. Don't quite like NamedLambdaExpression/NamedAggregateExpression. I added comments. We can discuss.

File: src/NRules/NRules.RuleModel/Builders/AggregateBuilder.cs (hunk 2 above)
Comment (a046) [reply src/NRules/NRules.RuleModel/Builders/AggregateBuilder.cs:14]: I did consider both Tuple and KVP, I think they are good too. Only issue is the names of the properties, which is a minor problem.

If using KVP, there's a lot of generic noise as there's no equivalent of Tuple.Create, unless I make one somewhere for our cases I guess? I think I'll have to retain the .Find/.FindSingle on IEnumerable<KVP> extension methods anyway.

Advantage of KVP is that places that currently create dictionaries can possibly keep working without changes assuming the KVP generic types match.

Comment (snikolayev) [reply src/NRules/NRules.RuleModel/Builders/AggregateBuilder.cs:14]: Yeah, the generics noise of KVP is annoying. But I think the fact that you can just continue passing a dictionary in, when you don't have duplicates, is helpful.

Comment (a046) [reply src/NRules/NRules.RuleModel/Builders/AggregateBuilder.cs:14]: Ok, agreed, i'll make all of these changes and add to this PR shortly. Using dictionaries where appropriate will cut down on the noise quite a bit.
